packages/state-transition/src/cache/epochTransitionCache.ts

```diff
@@ -200,21 +206,35 @@ export function beforeProcessEpoch(
   const validators = state.validators.getAllReadonlyValues();
   const validatorCount = validators.length;
 
+  // pre-fill with true (most validators are active)
+  const isActivePrevEpoch = new Array(validatorCount).fill(true);
+  const isActiveCurrEpoch = new Array(validatorCount).fill(true);
+  const isActiveNextEpoch = new Array(validatorCount).fill(true);
+
+  // During the epoch transition, additional data is precomputed to avoid traversing any state a second
+  // time. Attestations are a big part of this, and each validator has a "status" to represent its
+  // precomputed participation.
+  // - proposerIndex: number; // -1 when not included by any proposer
+  // - inclusionDelay: number;
+  // - flags: number; // bitfield of AttesterFlags
+  const proposerIndices = new Array<number>(validatorCount).fill(-1);
+  const inclusionDelays = new Array<number>(validatorCount).fill(0);
+  const flags = new Array<number>(validatorCount).fill(0);
```

Comment:
yes I see the improvement by prepopulate Arrays of `validatorCount`
but can we keep `AttesterStatus` data structure, ie 
```typescript
const defaultStatus = createAttesterStatus();
const statuses = new Array<AttesterStatus>(validatorCount).fill({...defaultStatus});
```
that way we'll access the array once instead of 3 if we decompose per loop

Reply:
you still have an "access" any time you access an object member.

For example:
```ts
const isActiveCurrEpoch = ... as boolean[];
const flags = ... as number[];
for (let i = 0; i < N; i++) {
  isActiveCurrEpoch = ...; // access 1
  flags[i] = ...; // access 2
}
```
vs
```ts
const statuses = ... as AttesterStatus[]
for (let i = 0; i < N; i++) {
  const status = statuses[i]; // access 1
  status.active = ...; // access 2
  status.flag = ...; // access 3
}
```

You can't avoid the accesses. But! using separate arrays will avoid one level of indirection.

Reply:
so I created the benchmark to compare AttesterStatus array vs decompose strategy (multiple arrays) [here](https://gist.github.com/twoeths/ff5b04f54dc7e641f2d21d5f0e95292d)
it shows that once you have an object, it does not matter how many properties you access. But accessing an item of an array does matter
```
  AttesterStatus vs decompose
    ✔ AttesterStatus - 1500000 4 properties                               703.9046 ops/s    1.420647 ms/op        -        143 runs  0.704 s
    ✔ decompose - 1500000 4 properties                                    475.0937 ops/s    2.104848 ms/op        -        146 runs  0.808 s
    ✔ AttesterStatus - 1500000 3 properties                               703.3605 ops/s    1.421746 ms/op        -        143 runs  0.704 s
    ✔ decompose - 1500000 3 properties                                    611.7626 ops/s    1.634621 ms/op        -        125 runs  0.704 s
    ✔ AttesterStatus - 1500000 2 properties                               699.6060 ops/s    1.429376 ms/op        -        284 runs  0.907 s
    ✔ decompose - 1500000 2 properties                                    713.0379 ops/s    1.402450 ms/op        -        218 runs  0.807 s
    ✔ AttesterStatus - 1500000 1 properties                               694.6475 ops/s    1.439579 ms/op        -        352 runs   1.01 s
    ✔ decompose - 1500000 1 properties                                    1065.190 ops/s    938.8000 us/op        -        324 runs  0.804 s

```
based on that benchmark, I analyze the access of AttesterStatus properties:

|file name | how many properties | property details|
|---------|---------------------|--------|
|validatorMonitors.ts|3|inclusionDelay, flag, isActiveCurrEpoch |
|getAttestationDeltas.ts|2|flags, proposerIndex|
|getRewardsAndPenalties.ts | 1 | flags|
|processInactivityUpdates.ts| 1 | flags|

based on the benchmark of 1.5M validators:
- using AttesterStatus array, it takes 1.42 * 4 = __5.68ms__
- using decompose arrays, it takes: 0.938 * (3 + 2 + 1 + 1) = 0.938 * 7 = __6.566ms__

so the decompose strategy is not better than using AttesterStatus array, even if it's an improvement, it's only about some `ms` better and it's not worth
I think the majority of the improvement comes from the preallocation of arrays, we could make it even better if we reuse them over multiple epoch transitions

Reply:
> so the decompose strategy is not better than using AttesterStatus array

Ok, I will try taking the latest commit's strategy but using AttesterStatuses rather than separate arrays.

> we could make it even better if we reuse them over multiple epoch transitions

Yes, the latest commit reuses arrays across epoch transitions. It shows even better results (see feat3 for the past 12 hrs or so.

Reply:
> I will try taking the latest commit's strategy but using AttesterStatuses

```ts
const statuses = new Array<AttesterStatus>(validatorCount).fill({...defaultStatus});
```
Unfortunately, this doesn't work, we won't be able to use the `fill` method.

Reply:
> Unfortunately, this doesn't work, we won't be able to use the fill method, this will create an array with all elements pointing to the same object.

could you try
```typescript
const statuses = new Array<AttesterStatus>(validatorCount).fill(() => createAttesterStatus());
```

this does not point all indices to the same object

Reply:
anyway I only expect to use the `fill()` method once if `statuses.length = 0` (the 1st time). From the 2nd time we only need to populate from the old length to new length to reuse `AttesterStatus` objects
```typescript
for (let i = oldLength; i < newLength; i++) {
  statuses[i] = createAttesterStatus();
}
```

Reply:
Yeah, unfortunately, flags, proposerIndex, and inclusionDelay have to be zeroed out for all old AttesterStatus objects too. Will try it out and see what perf looks like

Reply:
`cayman/prealloc-attester-statuses` branch now deployed to feat3

Reply:
> [cayman/prealloc-attester-statuses](https://github.com/ChainSafe/lodestar/compare/unstable...cayman/prealloc-attester-statuses) branch now deployed to feat3

metrics aren't looking as good as this branch... same memory usage, slower epoch transition time

this branch, before 14:00, cayman/prealloc-attester-statuses after 15:00
![Screenshot from 2024-07-13 14-03-17](https://github.com/user-attachments/assets/4566e5ea-f85f-4b26-8e8d-0c01c3bd430a)

Reply:
oh yes I see the clear difference. I guess memory allocation when array of objects grows vs array of basic types grows is the key different there
this is lesson learn for the next improvement @wemeetagain  👍